### PR TITLE
feat: Code block recording

### LINKED ISSUE #82: Code block recording
A mode for programmatic usage of AppMap would be useful to allow the developer to selectively record pieces of code.

While we can't instrument code that's already been loaded, we could have a mode that instruments the code normally but leaves the instrumentation disabled (eg. `appmap-node --inactive` or `--disabled`). The programmer could then use an exposed API to record snippets of interest, eg.

```typescript
const appmap = AppMap.record(() => {
  hello("world");
});
```

### PR DESCRIPTION
**Pull request by zermelo-wisen**
Fixes #82

### CODE REVIEW

**Review by dividedmind — CHANGES_REQUESTED** [review]
Block recording should work like this:
```js
import { record } from 'appmap-node';

const appmap = record(() => {
  hello("world");
});
```

Note you don't need to add a flag — just have a separate façade (NOT _bin.js_) for importing^Wrequiring (use _package.json_ to direct _require_/_import_ to the right file) and drop the process recording when it's loaded. You'll also need to expose the AppMap type so the clients can consume the results with confidence.

You probably don't need an extra _recordAsync_ flavor — just do async recording if the function is async and/or returns a promise.

File: src/Recording.ts [hunk 1]
```diff
@@ -212,6 +212,11 @@ export default class Recording {
   get running(): boolean {
     return !!this.stream;
   }
+
+  readAppMap(): AppMap.AppMap | undefined {
```
Comment (dividedmind) [inline src/Recording.ts:216]: When does this return `undefined`?

File: src/facade.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+disableGlobalRecording();
+
+export function record(block: () => unknown): AppMap.AppMap | undefined {
+  start(new Recording("block", "block", new Date().toISOString()));
+  startCodeBlockRecording();
+  try {
+    block();
+  } finally {
+    stopCodeBlockRecording();
+    recording.finish();
```
Comment (dividedmind) [inline src/facade.ts:22]: Should probably say `if (!recording.finish()) return`.

File: src/facade.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  start(new Recording("block", "block", new Date().toISOString()));
+  startCodeBlockRecording();
+  try {
+    await block();
+  } finally {
+    stopCodeBlockRecording();
+    recording.finish();
+  }
+  return recording.readAppMap();
+}
```
Comment (dividedmind) [inline src/facade.ts:37]: > My last decision was to make separate record functions for sync and async usages. I thought this was the best option for the client code. Otherwise the return type of the record function had to be `Promise<AppMap.AppMap | undefined> | AppMap.AppMap | undefined`, complicating consumption of the result by client code. Clients would need to make type assertions in Typescript to get intellisense etc. Now they can access members easilly: `appmap?.events.length`.

You can overload the declaration:

```typescript
type NotPromise<T> = T extends Promise<unknown> ? never : T;

export function record<T>(block: () => NotPromise<T>): AppMap.AppMap | undefined;
export function record<T>(block: () => Promise<T>): Promise<AppMap.AppMap> | undefined;

export function record(block: () => unknown): AppMap.AppMap | Promise<AppMap.AppMap> | undefined {
  start(new Recording("block", "block", new Date().toISOString()));
  startCodeBlockRecording();
  try {
    const result = block();
    if (result instanceof Promise) return result.then(finishRecording, finishRecording);
  } catch {
    // intentionally empty
  }
  return finishRecording();
}

function finishRecording(): AppMap.AppMap | undefined {
  stopCodeBlockRecording();
  if (!recording.finish()) return;
  return recording.readAppMap();
}
```

File: test/codeBlock/index.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+  const appmap1 = record(() => {
+    hello("world");
+  });
+
+  hello("123");
+
+  assert(appmap1.events.filter((e) => e.event == "call").length == 1);
+
+  const appmap2 = await recordAsync(() => {
```
Comment (dividedmind) [inline test/codeBlock/index.js:20]: This should probably be an async function, right?

File: test/codeBlock/package.json [hunk 5]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (dividedmind) [inline test/codeBlock/package.json:1]: Why is this needed?

File: src/facade.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  stopCodeBlockRecording,
+} from "./recorderControl";
+import Recording from "./Recording";
+
+// Since _this_ module is loaded, we'll do code block recording only.
+recording.abandon();
+disableGlobalRecording();
+
+export function record(block: () => unknown): AppMap.AppMap | undefined {
+  start(new Recording("block", "block", new Date().toISOString()));
```
Comment (dividedmind) [inline src/facade.ts:16]: The record function should probably check if the code is instrumented and throw otherwise, telling the user to run with `appmap-node`. Otherwise it will just mysteriously fail to produce anything. (Add a test case for that!)

**Review by dividedmind — CHANGES_REQUESTED** [review]
Can you please also split the two refactorings (*Appmap.d.ts* -> *AppMap.ts*, *emit()*) into separate commits with messages explaining the rationales? It'll make it much clearer what's going on when someone reads it down the line. Thanks!

File: src/Recording.ts [hunk 7]
```diff
@@ -172,6 +173,11 @@ export default class Recording {
   }
 
   private emit(event: unknown) {
+    // Check here if we should record instead of requiring each
+    // possible hook to check it.
+    // Checking this inside hooks too, will save CPU cycles only
+    // by preventing unnecessary event creation.
```
[the remark this answers is not in the document]
Comment (dividedmind) [reply src/Recording.ts]: I think it was already fine as it was — `record()` method already checked if `shouldRecord()`; I just thought it would be good to say that in this comment :)

[the remark this answers is not in the document]
Comment (zermelo-wisen) [reply src/Recording.ts]: Just to clarify, all the hooks which call event creation methods directly (sqlite, prisma, pg, mysql, mongo, http) will create the event objects, but the events will not be emitted when we should not record. I believe you meant these by some wasted events.

[the remark this answers is not in the document]
Comment (zermelo-wisen) [reply src/Recording.ts]: perf commit rolled back.

**Review by dividedmind — APPROVED** [review]
